raiden/network/transport/matrix/transport.py

```diff
@@ -1384,3 +1393,151 @@ def leave(room: Room):
                     f'node:{pex(self._raiden_service.address)} '
                     f'user_id:{self._user_id}'
                 )
+
+    def send_to_device(self, address: Address, message: Message) -> None:
+        """ Sends send-to-device events to a all known devices of a peer without retries. """
+        user_ids = self._address_mgr.get_userids_for_address(address)
+
+        data = {
+            user_id:
+                {'*': json.dumps(message.to_dict())}
+            for user_id in user_ids}
+
+        return self._client.api.send_to_device('m.to_device_message', data)
+
+    def _handle_to_device_message(self, event):
+        """
+        Handles to_device_message sent to us.
+        - validates peer_whitelisted
+        - validates userid_signature
+        Todo: Currently doesnt do anything but logging when a to device message is received.
+        """
+        sender_id = event['sender']
+
+        if (
+                event['type'] != 'm.to_device_message' or
+                self._stop_event.ready() or
+                sender_id == self._user_id
+        ):
+            # Ignore non-messages and our own messages
+            return False
+
+        user = self._get_user(sender_id)
+        peer_address = validate_userid_signature(user)
+        if not peer_address:
+            self.log.debug(
+                'To_device_message from invalid user displayName signature',
+                peer_user=user.user_id,
+            )
+            return False
+
+        # don't proceed if user isn't whitelisted (yet)
+        if not self._address_mgr.is_address_known(peer_address):
+            # user not start_health_check'ed
+            self.log.debug(
+                'ToDevice Message from non-whitelisted peer - ignoring',
+                sender=user,
+                sender_address=pex(peer_address),
+            )
+            return False
+
+        is_peer_reachable = self._address_mgr.get_address_reachability(peer_address) is (
+            AddressReachability.REACHABLE
+        )
+
+        if not is_peer_reachable:
+            self.log.debug('Forcing presence update', peer_address=peer_address, user_id=sender_id)
+            self._address_mgr.force_user_presence(user, UserPresence.ONLINE)
+            self._address_mgr.refresh_address_presence(peer_address)
+
+        data = event['content']
+        if not isinstance(data, str):
+            self.log.warning(
+                'Received ToDevice Message body not a string',
+                peer_user=user.user_id,
+                peer_address=to_checksum_address(peer_address),
+            )
+            return False
+
+        messages: List[Message] = list()
+
+        if data.startswith('0x'):
```

Comment:
Could this message parsing logic be extracted into an `utils` and deduplicated from handle message?

Reply:
Done